lib/iris/experimental/stratify.py

```diff
@@ -0,0 +1,179 @@
+            result.add_aux_coord(coord.copy(), dims)
+    for coord in cube.derived_coords:
+        dims = cube.coord_dims(coord)
+        if z_dim not in dims:
+            result.add_aux_coord(coord.copy(), dims)
+
+
+def relevel(phenom, level_data, target_levels, interp_direction=0,
+            interpolator=None):
+    """
+    Interpolate the given phenomenon onto the given levels of level_data.
+
+    For example, suppose we have two datasets `P(i,j,k)` and `H(i,j,k)`
+    and we want ``P(i,j,H)`` we call relevel with
+    ``phenom=P`` and ``level_data=H`` and ``target_levels`` being
+    an array of the values of ``H`` we would like.
+
+    This routine is especially useful for computing isosurfaces of
+    phenomenon that are generally monotonic in the direction of interpolation
+    (such as height/pressure and salinity/depth).
+
+    Parameters
+    ----------
+    phenom : :class:`iris.cube.Cube`
+        The data to be re-levelled.
+
+    level_data : :class:`~iris.cube.Cube` or :class:`~iris.coord.Coord`
+        The coordinate values (cube/coord) in the same system as
+        ``target levels``.
+        All dimensions of ``level_data`` must be broadcastable to ``phenom``
+
+    target_levels : array-like
+        The levels of ``level_data`` to pick from ``phenom``.
+        For the default interpolator, this must be a 1d array.
+        For a custom interpolator this may be multi-dimensional and must
+        broadcast to phenom and level_data except in the phenomenon's
+        ``interp_direction`` dimension.
+
+    interp_direction : int or :class:`~iris.coords.Coord` instance
+        The direction of interpolation. This is necessary because
+        level_data is rarely 1D. This is often the z dimension coordinate
+        of ``phenom``.
+
+    interpolator : callable or None
+        The interpolator to use when computing the interpolation. The function
+        will be passed the following positional arguments::
+
+            (target levels, level data, phenom data, interp axis)
+
+        If the interpolator is None, :func:`stratify.interpolate` will be used
+        with linear interpolation and NaN extrapolation.
+
+        An example of constructing an alternative interpolation scheme:
+
+            from functools import partial
+            interpolator = partial(stratify.interpolate,
+                                   interpolation=stratify.INTERPOLATE_NEAREST,
+                                   extrapolation=stratify.EXTRAPOLATE_LINEAR)
+
+    """
+    # Identify the z-coordinate within the phenom cube
+    if isinstance(interp_direction, Coord):
+        [interp_dim] = phenom.coord_dims(interp_direction)
+    else:
+        interp_dim = interp_direction
+
+    if isinstance(level_data, Coord):
+        source_data = level_data.points
+    else:
+        source_data = level_data.data
+
+    # phenom and level_data must be broadcastable.
+    phenom_data, source_data = np.broadcast_arrays(phenom.data,
```

Comment:
Just in case this has been missed, a common user might be using something like:
`source(z, y, x) -> target(z', y', x')`

That is:
```!python
src_z_coord = source.coord(axis='z')
relevel(source, src_z_coord,  target.coord(axis='z), src_z_coord)
```

This is going to throw a wobbler here at the broadcasting and I'm not 100% people will realise they simply need to transpose their source first i.e. `source(z, y, x) -> source(y, x, z)` to ensure they are broadcastable.